### ISSUE TITLE
Visual indication that notebook is still loading / cluster is being initialized

### ISSUE THREAD

**Issue by ryan-williams** [andypetrella/spark-notebook#294]
When I start an example notebook on YARN, it takes ~40s to initialize before even a simple command will return.

Sped-up screencast:

![](http://cl.ly/image/312q0K2E012D/sn-init2.gif)

Before I ran `1+2` in the notebook, the "kernel is idle" icon was showing; there's no indication that the notebook is still being initialized under the hood.

It would be nice if this was made clear in the UI.

In particular, I was bit by this combined with #293 because my notebooks were actually just never successfully initialized, and commands I typed would hang forever.

**Comment by ryan-williams** [comment]
NB: spark-notebook logs each notebook's driver stdout in its `logs/` directory, in a file with a name based on the randomly generated "session" name for your notebook, e.g. `logs/sn-session-b6e1205b-a194-481d-850c-efe8f6093c72-Untitled52.snb.log`.

I've found keeping a `tail -f logs/*Untitled52.snb.log` (where `Untitled52` is the name of my notebook) open in one terminal window while my notebook is running, and especially when it is initializing itself, to be super helpful; you can see whether/when it is connecting to YARN, creating a YARN app, etc.

**Comment by ryan-williams** [comment]
Just following up on this: it's really hard to figure out what notebooks are active in SN.

Right now, I look at the "Running" tab for a SN server used by several members of my team, and see 3 notebooks running:

![](http://cl.ly/image/2v080U3f2Z3N/Screen%20Shot%202015-09-12%20at%201.34.42%20PM.png)

However, I go to my YARN RM and there is only one running YARN app on our cluster:

![](http://cl.ly/image/0U1z0h2T3M0D/Screen%20Shot%202015-09-12%20at%201.35.56%20PM.png)

Which one of the three notebooks is that? Why aren't the other two actually connected to YARN apps?

I open each of the three running notebooks, go to to Edit > Edit Notebook Metadata, and see that they are all `"spark.master": "yarn-client"`. This is clumsy to check and has still not answered my question.

Finally I look in the `logs` folder of my server directory, see the last few files representing those notebooks, `less` them and look at the end to see what was going on.

Two of them, corresponding to "Untitled2" and "DREAM Challenge…" have exceptions that imply that the app has ended, e.g.:
```
org.apache.spark.SparkException: Yarn application has already ended! It might have been killed or unable to launch application master.
```

The third, "tim experimenting", has no obvious log file that has been recently touched:

```
$ lt logs
-rw-rw-r-- 1 willir31 hammej03a 728K Sep 11 22:04 sn-session-461facb9-df75-46fd-8f1a-ea9245458bc6-Untitled82.snb.log
-rw-rw-r-- 1 willir31 hammej03a  62K Sep 11 22:15 sn-session-3ae63d30-3ecd-442f-8765-4e32e31b19e2-Untitled83.snb.log
-rw-rw-r-- 1 willir31 hammej03a 651K Sep 11 22:15 sn-session-f1a82338-c7b6-4949-8cc4-6c7f2ddfa384-Untitled84.snb.log
-rw-rw-r-- 1 willir31 hammej03a  67K Sep 11 23:22 sn-session-58280f7e-7b77-4a49-9010-2b1541200820-PT189 Coverage.snb.log
-rw-rw-r-- 1 willir31 hammej03a 682K Sep 11 23:24 sn-session-7b510a64-a7ec-4bc4-a7ad-d1bca9a9c813-Untitled1.snb.log
-rw-r--r-- 1 willir31 hammej03a  32K Sep 12 13:25 sn-session-506eaa49-814d-4d3b-b301-85cd578158a2-Untitled2.snb.log
-rw-r--r-- 1 willir31 hammej03a 132K Sep 12 13:38 sn-session-8fc2fa89-09be-4e92-a4fb-e98bde7da7e4-Untitled3.snb.log
-rw-r--r-- 1 willir31 hammej03a  90K Sep 12 14:16 sn-session-84bd1fb0-cdbd-4cb2-a6aa-77d4cc05cde1-old\DREAM Challenge Synth3.snb.log
-rw-rw-r-- 1 willir31 hammej03a 656K Sep 12 17:37 application.log
```

It turns out that "Untitled3" is the log file for that notebook, which has since been renamed, and inside it I can see that it corresponds to the running YARN app, `application_1439224376754_6222`.

Clearly, this all could be made so much easier! Let the notebooks display info about their YARN app, and maybe links to their logs as well, to avoid the rename-induced log-file-name-mismatch I noted at the end there.

**Comment by andypetrella** [comment]
MMMmmmmh I see the pain points. So, I see there are three main things we need to fix:
* creating a notebook should before ask for a name → so this problem will at least occur less often
* see how the logs file can also change after the notebook has been renamed (probably a filesystem `mv`, don't expect that easily doable with slf4j...)
* see how we can get some feedback on connection status/progress on the cluster

Changing the name is Yarn is trickier because will probably need to restart the app, I don't think it is possible to change an app name in either cluster types (yarn, mesos, standalone, ...)

It's quite some work, but we can plan something on that.
WDYT?

**Comment by ryan-williams** [comment]
Yes, good ideas. Two things I think you should **not** do, because they seem hard (as you've noted):

* hot-renaming log files
* changing the name of YARN apps

The easier/better solution here is to just let me get the info I need from the browser:

* let me see a notebook's log file in the browser, linked from the notebook.
  * similar to how YARN makes it pretty easy to get to container std{out,err} from the YARN RM web UI.
* let me see the YARN app ID in the notebook, and link to YARN RM web UI for that app to check status.
  * bonus: provide some indication of the status in notebook web UI itself.

**Comment by andypetrella** [comment]
cf #321
